### ISSUE TITLE
knative mirror job not working ?

### ISSUE THREAD

**Issue by matzew** [openshift/release#3834]
/kind question

It seems that we have no working image mirror job for our knative images.
They are configured here:

* https://github.com/openshift/release/blob/master/cluster/ci/jobs/image-mirror.yaml#L331-L377

and the "mapping" file is here:

* https://github.com/openshift/release/blob/master/cluster/ci/config/mirroring/knative/knative_v0_6_quay

Is it possible that we lack some hook here:

* https://github.com/openshift/release/blob/master/Makefile#L272

## What did you expect to happen?

After the CI builds are done, I'd expect that the images for the `0.6.0` would be pushed to quay.io.

But they are not there:

## How can the error be reproduced?

Trigger a CI build against the 0.6.0 branch of one of the knative forks on openshift

**Comment by matzew** [comment]
@petr-muller do you have an idea ?

**Comment by matzew** [comment]
the images are actually in the `registry`.

For instance:

```
docker pull registry.svc.ci.openshift.org/openshift/knative-v0.6.0:knative-eventing-pong
```

gives me:
```
Trying to pull repository registry.svc.ci.openshift.org/openshift/knative-v0.6.0 ... 
sha256:9bec768d51ffd6e9c4c9bde4ab810bed040a489e5aca274c0e4d34b481e97e10: Pulling from registry.svc.ci.openshift.org/openshift/knative-v0.6.0
5dfdb3f0bcc0: Pull complete 
99f178453a43: Pull complete 
93f49b19cd6f: Pull complete 
d04a6e71e492: Pull complete 
afe138467dd3: Pull complete 
Digest: sha256:9bec768d51ffd6e9c4c9bde4ab810bed040a489e5aca274c0e4d34b481e97e10
Status: Downloaded newer image for registry.svc.ci.openshift.org/openshift/knative-v0.6.0:knative-eventing-pong
```

**Comment by matzew** [comment]
Do we need something in here, for the knative bits:
https://github.com/openshift/release/blob/master/Makefile#L272-L274

**Comment by petr-muller** [comment]
I don't think you need anything around https://github.com/openshift/release/blob/master/Makefile#L272. I'm looking at the cluster and I do not even see `image-mirror-knative-v0.6-quay` CronJob there (there are ones for 0.4 and 0.5 though). So the reason why the images do not get publishes is that the cronjob does not even exist.

Our postsubmit calls the `postsubmit-update` make target, but this target does not seem to include the `image-mirror-setup` target which actually creates the job.

**Comment by petr-muller** [comment]
/reopen

Merging https://github.com/openshift/release/pull/3843 created the cronjob but it seems to be failing, I do not know why yet.

**Comment by stevekuznetsov** [comment]
Why are the CronJobs not in separate namespaces like we originally intended? I want the knative team to be able to debug this end-to-end

**Comment by matzew** [comment]
Ok, what's needed in order to change that - I think I do lack understanding of what actually needs to be done in order to fix it

**Comment by petr-muller** [comment]
So the reason whe CronJobs are failing is likely this:

1. The jobs [need](https://github.com/openshift/release/blob/master/cluster/ci/jobs/image-mirror.yaml#L376) a `/etc/imagemirror/knative/knative_v0_6_quay` file
2. `/etc/imagemirror` is a [mount](https://github.com/openshift/release/blob/master/cluster/ci/jobs/image-mirror.yaml#L367) of the [image-mirror](https://github.com/openshift/release/blob/master/cluster/ci/jobs/image-mirror.yaml#L35) ConfigMap
3. This ConfigMap is created in the [Makefile](https://github.com/openshift/release/blob/master/Makefile#L282) by `oc create configmap image-mirror --from-file=cluster/ci/config/mirroring/`
4. Your knative files are in a `knative` subdirectory of that directory, and `create configmap --from-file` is not recursive, so your files are not present in the CM

I think the good solution immediately is to create a separate CM from the `knative` subdir and point the jobs to it. Long-term you should really have a separate namespace (like @stevekuznetsov said) so that you can maintain and debug these bits yourself.

**Comment by matzew** [comment]
OK, will take a look - thanks!

**Comment by matzew** [comment]
@petr-muller is there more to it?  I did not see a push this night  

that said: we had to (before) push some images manually... but IMO that should not make a difference

**Comment by petr-muller** [comment]
@matzew Looks like you still pass `-f=/etc/imagemirror/knative/knative_v0_6_quay` to the `oc image mirror` command. `/etc/imagemirror` is a mount of the `knative-image-mirror` CM, so its items will be in that directory, not in the `knative` subdir:

```
            volumes:
            - name: config
              configMap:
                name: knative-image-mirror
...
            containers:
...
              volumeMounts:
              - name: config
                mountPath: /etc/imagemirror
...
```

So either fix the path of mount the CM as `/etc/imagemirror/knative`.

FWIW, I did an experiment and instantiated a CM from the fixed-up version of `image-mirror-knative-v0.6-quay` - it at least executed, but failed with the following:

```
error: unable to retrieve source image registry.svc.ci.openshift.org/openshift/knative-v0.6.0 by tag knative-build-webhook: manifest unknown: manifest unknown
error: unable to retrieve source image registry.svc.ci.openshift.org/openshift/knative-v0.6.0 by tag knative-build-creds-init: manifest unknown: manifest unknown
error: unable to retrieve source image registry.svc.ci.openshift.org/openshift/knative-v0.6.0 by tag knative-build-git-init: manifest unknown: manifest unknown
error: unable to retrieve source image registry.svc.ci.openshift.org/openshift/knative-v0.6.0 by tag knative-build-controller: manifest unknown: manifest unknown
error: unable to retrieve source image registry.svc.ci.openshift.org/openshift/knative-v0.6.0 by tag knative-build-nop: manifest unknown: manifest unknown
error: an error occurred during planning 
```

Not sure how to fix these.
